File: python/linked-list/linked_list.py
class Node(object):
    def __init__(self, value, next=None, prev=None):
        self.value = value
        self.next = next
        self.prev = prev

class LinkedList(object):
    def __init__(self):
        self.head = None
        self.tail = None

    def push(self, value):
        if self.head == None:
            self.head = Node(value)
            self.tail = self.head
        else:
            self.tail.next = Node(value, None, self.tail)
            self.tail = self.tail.next

    def unshift(self, value):
        if self.head == None:
            self.head = Node(value)
            self.tail = self.head
        else:
            self.head.prev = Node(value, self.head, None)
            self.head = self.head.prev

    def pop(self):
        value = self.tail.value
        if self.tail.prev:
            self.tail = self.tail.prev
            self.tail.next = None
        else:
            self.head = None
            self.tail = self.head
        return value
    
    def shift(self):
        value = self.head.value
        if self.head.next:
            self.head = self.head.next
            self.head.prev = None
        else:
            self.head = None
            self.tail = self.head
        return value
    
    def __len__(self):
        length = 0
        curr = self.head
        while curr:
            length += 1
            curr = curr.next
        return length

    def __iter__(self):
        self.n = None
        return self
    
    def __next__(self):
        if self.n == None:
            self.n = self.head
        else:
            self.n = self.n.next
        return self.n.value

File: typescript/linked-list/linked-list.ts
class Node<T> {
  value: T
  prev: Node<T> | null
  next: Node<T> | null

  constructor(value: T, prev: Node<T> | null, next: Node<T> | null) {
    this.value = value
    this.prev = prev
    this.next = next
  }
}

class LinkedList<T> {
  head: Node<T> | null= null
  tail: Node<T> | null = null
  _count = 0
  
  push(value: T): void {
    const newNode = new Node(value, this.tail,  null)
    if (this.tail) {
      this.tail.next = newNode
    }
    this.tail = newNode
    this.head = this.head || this.tail
  }

  pop(): T | null {
    const res = this.tail
    this.tail = this.tail ? this.tail.prev : this.head
    this.head = this.tail ? this.head : null
    return res ? res.value : null
  }

  unshift(value: T): void {
      const newNode = new Node(value, null, this.head)
      if (this.head) {
        this.head.prev = newNode
      }
      this.head = newNode
      this.tail = this.tail || this.head
  }

  shift(): T | null {
    const res = this.head
    this.head = this.head ? this.head.next : this.tail
    this.tail = this.head ? this.tail : null
    return res ? res.value : null
  }

  find(value: T): Node<T> | null {
    let curr = this.head
    while (curr && curr.value !== value) {
      curr = curr.next
    }
    return curr
  }

  remove(node: Node<T>): void {
    if (node.next) {
      node.next.prev = node.prev
    }
    if (node.prev) {
      node.prev.next = node.next
    }
    this.head = node === this.head ? node.next : this.head
    this.tail = node === this.tail ? node.prev : this.tail
  }

  delete(value: T): void {
    const nodeToDelete = this.find(value)
    if (nodeToDelete) {
      this.remove(nodeToDelete)
    }
  }

  count = (): number => {
    let count = 0
    for (let curr = this.head; curr; curr = curr.next, count++) {}
    return count
  }
}

export default LinkedList
